[Merged by Bors] - feat: codimension two subspaces have path-connected complement

grunweg opened the pull request:
From sphere-eversion (not written by me).
This basically shows that subspaces of codimension at least two are ample sets (PRed in #11342).

---------

In sphere-eversion, the hypotheses was spelled as "2 \leq Module.rank E"; I have changed this to "1 < Module.rank E" to match the remaining file. (The latter seems to be used more in mathlib, but either is pretty rare.)

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

sgouezel commented on Mathlib/Analysis/Convex/Normed.lean:
```diff
@@ -155,3 +153,87 @@ theorem isConnected_setOf_sameRay_and_ne_zero {x : E} (hx : x ≠ 0) :
+
+/-- In a real vector space whose dimension is at least two,
+the complement of `{0}` is path-connected. -/
+theorem isPathConnected_compl_zero_of_two_le_dim (hdim : 2 ≤ Module.rank ℝ E) :
```
We already have [isPathConnected_compl_singleton_of_one_lt_rank](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/NormedSpace/Connected.html#isPathConnected_compl_singleton_of_one_lt_rank)

grunweg replied:
Well spotted; addressed now.

grunweg commented:
Rebased on master (sorry for the force-push; only the last commit is new since the last review comment).

grunweg commented:
Bumped heartbeats in `CondexpL1.lean`; I have no idea why my change pushed things over the edge.

grunweg commented:
Good point; I agree. I have removed these lemmas and moved `isPathConnected_compl_of_two_le_codim` to `Analysis.NormedSpace.Connected` as well. I am not convinced if `Submodule.connectedComponentIn_eq_self_of_two_le_codim` is carrying its weight now; I can certainly remove it now and reconsider in the dependent PRs.
awaiting-review

grunweg commented:
Wait, let me clean up the diff first.

grunweg commented:
Now it's good for another look.
awaiting-review

sgouezel commented on Mathlib/Analysis/NormedSpace/Connected.lean:
```diff
@@ -168,3 +168,30 @@ theorem isPreconnected_sphere (h : 1 < Module.rank ℝ E) (x : E) (r : ℝ) :
+
+section
+
+variable {E F : Type*} [AddCommGroup F] [Module ℝ F] [TopologicalSpace F]
+  [AddCommGroup E] [Module ℝ E] [TopologicalSpace E] [TopologicalAddGroup F] [ContinuousSMul ℝ F]
```
Do you really use `E`? Below it seems it's only a submodule.

grunweg commented:
awaiting-review

sgouezel commented:
bors r+
Thanks!

grunweg commented:
Thank you for the reviews!
